### PR TITLE
src/redist/project.json

### PR DESCRIPTION
**Pull request by piotrpMSFT**
Creates a redist/project.json project that pulls together all executables that make up the CLI
- To enable this, makes each tool [nuget, msbuild, csc, etc.] a standalone project
- Removes unnecessary dependencies from dotnet/project.json, which was turning into a dependency dumping ground
- Makes nuget run out-of-proc from dotnet.dll

### CODE REVIEW

File: build/Microsoft.DotNet.Cli.Compile.targets [hunk 1]
```diff
@@ -208,12 +210,8 @@
       </ItemGroup>
 
       <Copy SourceFiles="@(MSBuildTargetsToCopy)"
-            DestinationFiles="@(MSBuildTargetsToCopy->'$(SdkOutputDirectory)\%(RecursiveDir)%(Filename)%(Extension)')" />
+            DestinationFiles="@(MSBuildTargetsToCopy->'$(SdkOutputDirectory)/%(RecursiveDir)%(Filename)%(Extension)')" />
 
-    <!-- Temporary workaround for MSBuild placing their .props files in the old 14.1 directory -->
-    <Move SourceFiles="$(SdkOutputDirectory)/14.1/Microsoft.Common.props"
```
Comment (eerhardt) [inline build/Microsoft.DotNet.Cli.Compile.targets]: I think you still need this workaround

File: src/dotnet/commands/dotnet-restore/NuGetForwardingApp.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+                GetNuGetExePath(),
+                argsToForward);
+        }
+
+        public int Execute()
+        {
+            return _forwardingApp.Execute();
+        }
+
+        private static string GetHostPath()
```
Comment (eerhardt) [inline src/dotnet/commands/dotnet-restore/NuGetForwardingApp.cs]: Is this method used?

File: src/nuget/project.json [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+      "version": "3.5.0-rc1-1697",
+      "exclude": "compile"
+    },
+    "NuGet.CommandLine.XPlat": "3.5.0-rc1-1697"
+  },
+  "frameworks": {
+    "netcoreapp1.0": {
+      "imports": [
+        "dnxcore50",
+        "netstandardapp1.5",
```
Comment (eerhardt) [inline src/nuget/project.json]: Are all these imports still needed?  I thought NuGet fixed up all their packages to not require imports anymore.

File: src/nuget/project.json [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  "version": "1.0.0-featmsbuild-*",
+  "buildOptions": {
+    "emitEntryPoint": true
+  },
+  "dependencies": {
+    "Microsoft.NETCore.App": {
+      "type": "platform",
+      "version": "1.0.0"
+    },
+    "NuGet.Commands": {
```
Comment (eerhardt) [inline src/nuget/project.json]: NuGet.CommandLine.XPlat references this.  Why do we need an explicit reference to it?

File: src/nuget/project.json [hunk 5]
```diff
@@ -0,0 +1,27 @@
+{
```
Comment (eerhardt) [inline src/nuget/project.json]: Should this directory be "tool_nuget" to align with "tool_msbuild"

File: src/dotnet/commands/dotnet-restore/NuGetForwardingApp.cs [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+{
+    public class NuGetForwardingApp
+    {
+        private const string s_nugetExeName = "NuGet.CommandLine.XPlat.dll";
+        private readonly ForwardingApp _forwardingApp;
+
+        public NuGetForwardingApp(string[] argsToForward)
+        {
+            _forwardingApp = new ForwardingApp(
+                GetNuGetExePath(),
```
Comment (eerhardt) [inline src/dotnet/commands/dotnet-restore/NuGetForwardingApp.cs]: What deps file and runtimeconfig are used here?

File: src/nuget/project.json [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+  "version": "1.0.0-featmsbuild-*",
```
Comment (eerhardt) [inline src/nuget/project.json]: With this new project, are we getting a new .dll named "nuget.dll" in the output folder?  if yes, is that intentional?

File: build/Microsoft.DotNet.Cli.Compile.targets [hunk 8]
```diff
@@ -121,36 +129,26 @@
...
 
-      <RemoveAssetFromDepsPackages DepsFile="$(SdkOutputDirectory)/csc.deps.json"
-                                   SectionName="runtimeTargets"
-                                   AssetPath="$(BinaryToCorehostifyRelDir)/%(BinariesToRemove.Identity).exe" />
-      <RemoveAssetFromDepsPackages DepsFile="$(SdkOutputDirectory)/dotnet.deps.json"
-                                   SectionName="runtimeTargets"
-                                   AssetPath="$(BinaryToCorehostifyRelDir)/%(BinariesToRemove.Identity).exe" />
-
-      <!-- cleanup compilers project output we don't need -->
-      <Delete Files="$(SdkOutputDirectory)/compilers%(PublishOutputExtensions.Identity)" />
```
Comment (eerhardt) [inline build/Microsoft.DotNet.Cli.Compile.targets]: Do we still need to delete the compilers output?  How about the src\nuget output?

File: src/dotnet/commands/dotnet-build3/MSBuildForwardingApp.cs [hunk 9]
```diff
@@ -56,19 +54,18 @@ private static string GetHostPath()
 
         private static string GetRuntimeConfig()
```
Comment (eerhardt) [inline src/dotnet/commands/dotnet-build3/MSBuildForwardingApp.cs]: This and GetDepsFile don't appear to be called anymore.

**Comment by eerhardt** [comment]
>     public class Program

Is this file still needed?  We don't have it in the nuget, nor 'tool_msbuild'

---
Refers to: src/csc/Program.cs:5 in 3a55a41. [](commit_id = 3a55a4182f49c6c63072ea7b47f8efde2cdd7355, deletion_comment = False)

File: build/Microsoft.DotNet.Cli.Compile.targets [hunk 10]
```diff
@@ -3,8 +3,16 @@
   <Import Project="compile/Microsoft.DotNet.Cli.LzmaArchive.targets" />
   <Import Project="crossgen/Microsoft.DotNet.Cli.Crossgen.targets" />
 
+  <PropertyGroup>
+    <BinaryToCorehostifyRelDir>runtimes/any/native</BinaryToCorehostifyRelDir>
+    <CoreSDKDir>$(RepoRoot)/resources/core-sdk</CoreSDKDir>
```
Comment (eerhardt) [inline build/Microsoft.DotNet.Cli.Compile.targets]: not sure I am crazy about 'core-sdk' as a name for this directory.

How about `$(RepoRoot)/resources/MSBuild`  or `$(RepoRoot)/resources/MSBuildImports`.

Note that our .props and .targets are getting deleted from our repo now that they live in the dotnet/sdk repo.

File: src/redist/project.json [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+  },
+  "dependencies": {
+    "Microsoft.NETCore.App": {
+      "type": "platform",
+      "version": "1.0.0"
+    },
+    "csc": "1.0.0-featmsbuild-*",
+    "dotnet": "1.0.0-featmsbuild-*",
+    "nuget": "1.0.0-featmsbuild-*",    
+    "tool_msbuild": "1.0.0-featmsbuild-*"
```
Comment (brthor) [inline src/redist/project.json]: Should these have `"target": "project"`

File: build/Microsoft.DotNet.Cli.Compile.targets [hunk 12]
```diff
@@ -76,6 +83,7 @@
         <BinariesToRemove Remove="*" />
         <BinariesToRemove Include="csc" />
         <BinariesToRemove Include="vbc" />
+        <BinariesToRemove Include="MSBuild" />
```
Comment (eerhardt) [inline build/Microsoft.DotNet.Cli.Compile.targets]: Is it weird that this Item is called "BinariesToRemove", but I don't see them being removed?

File: src/dotnet/commands/dotnet-restore/NuGetForwardingApp.cs (hunk 6 above)
Comment (piotrpMSFT) [inline src/dotnet/commands/dotnet-restore/NuGetForwardingApp.cs]: build scripts duplicate dotnet.deps & dotnet.runtimeconfig for each tool we package with the sdk. This makes them independently runnable so e.g. msbuild doesn't need to understand the workarounds we have in place in CLI.

File: src/dotnet/commands/dotnet-restore/NuGetForwardingApp.cs (hunk 2 above)
Comment (piotrpMSFT) [inline src/dotnet/commands/dotnet-restore/NuGetForwardingApp.cs]: fixed.

File: src/nuget/project.json (hunk 5 above)
Comment (piotrpMSFT) [inline src/nuget/project.json]: fixed. tool_csc as well.

File: src/nuget/project.json (hunk 4 above)
Comment (piotrpMSFT) [inline src/nuget/project.json]: removed.

File: src/nuget/project.json (hunk 3 above)
Comment (piotrpMSFT) [inline src/nuget/project.json]: removed.

File: build/Microsoft.DotNet.Cli.Compile.targets (hunk 1 above)
Comment (piotrpMSFT) [inline build/Microsoft.DotNet.Cli.Compile.targets]: fixed. Also now deleting the 14.1 directory.

File: src/redist/project.json (hunk 11 above)
Comment (piotrpMSFT) [inline src/redist/project.json]: that's implicit.

File: build/Microsoft.DotNet.Cli.Compile.targets (hunk 8 above)
Comment (piotrpMSFT) [inline build/Microsoft.DotNet.Cli.Compile.targets]: No. we only publish redist, which transitively brings along the rest.

src/tool_* does not have a Program.cs so no binary gets generated, so nothing to clean up there either.

File: build/Microsoft.DotNet.Cli.Compile.targets (hunk 10 above)
Comment (piotrpMSFT) [inline build/Microsoft.DotNet.Cli.Compile.targets]: sounds good. When this stuff lived under src/dotnet its intent was really unclear. I changed to MSBuildImports.

File: build/Microsoft.DotNet.Cli.Compile.targets (hunk 12 above)
Comment (piotrpMSFT) [inline build/Microsoft.DotNet.Cli.Compile.targets]: They are removed from deps.json files.. I renamed to AssetsToRemoveFromDeps

File: src/dotnet/commands/dotnet-build3/MSBuildForwardingApp.cs (hunk 9 above)
Comment (piotrpMSFT) [inline src/dotnet/commands/dotnet-build3/MSBuildForwardingApp.cs]: I went through the file and removed all dead code.

File: src/nuget/project.json (hunk 7 above)
Comment (piotrpMSFT) [inline src/nuget/project.json]: no, we're not. no .cs files.

File: src/redist/project.json (hunk 11 above)
Comment (brthor) [inline src/redist/project.json]: Is it guaranteed implicitly?

IIRC we've run into issues where packages were being pulled where we expected projects. Being explicit here is better IMO.

File: build/Microsoft.DotNet.Cli.Compile.targets [hunk 13]
```diff
@@ -3,8 +3,16 @@
   <Import Project="compile/Microsoft.DotNet.Cli.LzmaArchive.targets" />
   <Import Project="crossgen/Microsoft.DotNet.Cli.Crossgen.targets" />
 
+  <PropertyGroup>
+    <BinaryToCorehostifyRelDir>runtimes/any/native</BinaryToCorehostifyRelDir>
+    <CoreSDKDir>$(RepoRoot)/resources/MSBuildImports</CoreSDKDir>
```
Comment (eerhardt) [inline build/Microsoft.DotNet.Cli.Compile.targets]: One last nit about this.  The $(CoreSDKDir) property is now out of date, and doesn't align with the folder it is referencing.

Comment (eerhardt) [inline build/Microsoft.DotNet.Cli.Compile.targets]: Same for $(CoreSDKContent) below.

---
In reply to: [74152883](https://github.com/dotnet/cli/pull/4024#discussion_r74152883) [](ancestors = 74152883)

File: src/redist/project.json [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+    "Microsoft.NETCore.App": {
+      "type": "platform",
+      "version": "1.0.0"
+    },
+    "tool_csc": "1.0.0-featmsbuild-*",
+    "dotnet": "1.0.0-featmsbuild-*",
+    "tool_nuget": "1.0.0-featmsbuild-*",
+    "tool_msbuild": "1.0.0-featmsbuild-*",
+
+    "MSBuild": "0.1.0-preview-00029-160805",
```
Comment (eerhardt) [inline src/redist/project.json]: Shouldn't all these MSBuild dependencies come through the "tool_msbuild" dependency?  Why duplicate them here?

**Comment by eerhardt** [comment]
:shipit:

**Pull request merged**